docs/DesignDocs/TEECallsRefactoring.md

```diff
@@ -0,0 +1,596 @@
+ * the generated code for internal use. */
+#include "core_t.c"
+```
+
+In addition, `core_t_wrapper.c` implements a function as the following code snippet that
+registers the table to an internal data structure.
+This function is invoked by the `oecore` library during the enclave initialization.
+With this, the enclave internally maintains multiple ECALL tables according to the number
+of EDL files.
+
+```
+/* Registers the core ECALL function table. */
+oe_result_t oe_register_core_ecall_function_table(void)
+{
+    const uint64_t table_id = OE_CORE_ECALL_FUNCTION_TABLE_ID;
+    const oe_ecall_func_t* ecalls = __oe_core_ecalls_table;
+    const size_t num_ecalls = __oe_core_ecalls_table_size;
+
+    return oe_register_ecall_function_table(table_id, ecalls, num_ecalls);
+}
+```
+
+**Wrapper of `core_u.c`**
+
+The goal of the wrapper of `core_u.c`, or `core_u_wrapper.c`, is allowing the
+ECALL stub code to pass the identifier of the ECALL table when making the ECALL.
+To this end, `core_u_wrapper.c` replaces the `oe_call_host_function()` function
+with the following code snippet.
+
+```
+/* Override oe_call_enclave_function() with _call_core_enclave_function(). */
+#define oe_call_enclave_function _call_core_enclave_function
+
+/* The ocall edge routines will use this function to route ecalls. */
+static oe_result_t _call_core_enclave_function(
+    oe_enclave_t* enclave,
+    uint32_t function_id,
+    const void* input_buffer,
+    size_t input_buffer_size,
+    void* output_buffer,
+    size_t output_buffer_size,
+    size_t* output_bytes_written)
+{
+    return oe_call_enclave_function_by_table_id(
+        enclave,
+        OE_CORE_ECALL_FUNCTION_TABLE_ID,
+        function_id,
+        input_buffer,
+        input_buffer_size,
+        output_buffer,
+        output_buffer_size,
+        output_bytes_written);
+}
+```
+
+Both the host and the enclave share the definition of the `OE_CORE_ECALL_FUNCTION_TABLE_ID`,
+which is a hard-coded value that represents the identifier of the table.
+Upon receiving the request of the ECALL, `oe_handle_call_enclave_function()` locates
+the table based on the table id and looks up the ECALL based on the function id
+within the table. See the following code snippet from the `oe_handle_call_enclave_function()`
+function.
+
+```
+// Resolve which ecall table to use.
+if (args_ptr->table_id == OE_UINT64_MAX)
+{
+    ecall_table.ecalls = __oe_ecalls_table;
+    ecall_table.num_ecalls = __oe_ecalls_table_size;
+}
+else
+{
+    if (args_ptr->table_id >= OE_MAX_ECALL_TABLES)
+        OE_RAISE(OE_NOT_FOUND);
+
+    ecall_table.ecalls = _ecall_tables[args_ptr->table_id].ecalls;
+    ecall_table.num_ecalls = _ecall_tables[args_ptr->table_id].num_ecalls;
+
+    if (!ecall_table.ecalls)
+        OE_RAISE(OE_NOT_FOUND);
+}
+```
+
+Limitation
+---------
+
+The current implementations pose a limitation to the scenario where the
+host instantiates different enclaves that use the same set of ECALLs
```

Comment:
Given the problem, there does not seem to be much value in the description of the ocall dispatching mechanism.

Reply:
Removed the ocall section.